feat: clean up outdated images from DOCR

naomi-lgbt opened the pull request:
Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [X] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [X] My pull request has a [descriptive title](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request?id=prepare-a-good-pr-title) (**not** a vague title like `Update index.md`)

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

Related #705 

<!-- Feel free to add any additional description of changes below this line -->

This still needs some testing (which is difficult to do when there aren't any outdated tags to delete).

Commit:
feat: clean up outdated digitalocean images

naomi-lgbt commented on .github/workflows/deploy-i18n.yml:
```diff
@@ -134,10 +134,54 @@ jobs:
           docker push registry.digitalocean.com/${{ secrets.DOCR_NAME }}/${{ matrix.site_tlds }}/news-${{ matrix.languages }}:latest
 
+      - name: Clean old DO images
+        uses: actions/github-script
```
I would have written this in bash, but parsing the JSON into a bash array was becoming an absolute nightmare.

raisedadead replied:
Have you used `jq`? It's going to change the way you look at json forever. https://jqplay.org

raisedadead replied:
<img width="823" alt="image" src="https://github.com/freeCodeCamp/news/assets/1884376/b56b8c1d-d2d5-4f8d-a5f3-5e8a8d410a6d">


Here is the filter:

```txt
doctl registry repository list-tags org/news-chinese --output=json | jq '[.[] | {updated_at, tag}]'
```

I typically cheat these days by asking ChatGPT the exact syntax instead of writing it by hand.

naomi-lgbt replied:
Life-saver. I've pushed a bash version.

naomi-lgbt commented on .github/workflows/deploy-i18n.yml:
```diff
@@ -134,10 +134,54 @@ jobs:
         if: ${{ always() }} # To ensure this step runs even when earlier steps fail
 
+  clean:
+    name: Clean up DO images
```
Because garbage collection runs at the registry level, this needs to run after all of the tags have been cleaned (rather than running once on each lang).

Additionally, because it sets the registry to read-only mode until the GC is complete, it needs to run after all of the new images are pushed.

raisedadead replied:
Is it possible to move the step `Clean old DO images` into this job?

Commit:
fix: use bash

Commit:
chore: run prettier

Commit:
fix(tools): delete every two days, not week.

naomi-lgbt commented:
I ran the script locally with some extra logging (and skipping the actual delete call):

```console
Found 28 tags.
Checking tag: latest
Updated: 2023-08-28T18:22:38Z\nUpdated Date: 1693246958\nNow: 1693248937\nDiff: 1979
Checking tag: e6f0942-20230828-1822
Updated: 2023-08-28T18:22:38Z\nUpdated Date: 1693246958\nNow: 1693248937\nDiff: 1979
Checking tag: e6f0942-20230828-1232
Updated: 2023-08-28T12:32:47Z\nUpdated Date: 1693225967\nNow: 1693248937\nDiff: 22970
Checking tag: e6f0942-20230828-1122
Updated: 2023-08-28T11:23:02Z\nUpdated Date: 1693221782\nNow: 1693248937\nDiff: 27155
Checking tag: e6f0942-20230827-1013
Updated: 2023-08-27T10:13:42Z\nUpdated Date: 1693131222\nNow: 1693248937\nDiff: 117715
Checking tag: e6f0942-20230827-0621
Updated: 2023-08-27T06:22:27Z\nUpdated Date: 1693117347\nNow: 1693248937\nDiff: 131590
Checking tag: e6f0942-20230827-0108
Updated: 2023-08-27T01:08:52Z\nUpdated Date: 1693098532\nNow: 1693248937\nDiff: 150405
Checking tag: e6f0942-20230826-1819
Updated: 2023-08-26T18:20:28Z\nUpdated Date: 1693074028\nNow: 1693248937\nDiff: 174909
Checking tag: e6f0942-20230826-1227
Updated: 2023-08-26T12:28:02Z\nUpdated Date: 1693052882\nNow: 1693248937\nDiff: 196055
Checking tag: 96a3e4c-20230826-0620
Updated: 2023-08-26T06:21:17Z\nUpdated Date: 1693030877\nNow: 1693248937\nDiff: 218060
Checking tag: 71b7ac2-20230826-0103
Updated: 2023-08-26T01:03:52Z\nUpdated Date: 1693011832\nNow: 1693248937\nDiff: 237105
Checking tag: 71b7ac2-20230825-1821
Updated: 2023-08-25T18:21:52Z\nUpdated Date: 1692987712\nNow: 1693248937\nDiff: 261225
Checking tag: 71b7ac2-20230825-1231
Updated: 2023-08-25T12:31:37Z\nUpdated Date: 1692966697\nNow: 1693248937\nDiff: 282240
Checking tag: 71b7ac2-20230825-0621
Updated: 2023-08-25T06:22:02Z\nUpdated Date: 1692944522\nNow: 1693248938\nDiff: 304416
Checking tag: 71b7ac2-20230825-0105
Updated: 2023-08-25T01:05:47Z\nUpdated Date: 1692925547\nNow: 1693248938\nDiff: 323391
Checking tag: 71b7ac2-20230824-1821
Updated: 2023-08-24T18:22:12Z\nUpdated Date: 1692901332\nNow: 1693248938\nDiff: 347606
Checking tag: 71b7ac2-20230824-1231
Updated: 2023-08-24T12:31:37Z\nUpdated Date: 1692880297\nNow: 1693248938\nDiff: 368641
Checking tag: 71b7ac2-20230824-0621
Updated: 2023-08-24T06:22:02Z\nUpdated Date: 1692858122\nNow: 1693248938\nDiff: 390816
Checking tag: 71b7ac2-20230824-0104
Updated: 2023-08-24T01:05:27Z\nUpdated Date: 1692839127\nNow: 1693248938\nDiff: 409811
Checking tag: 71b7ac2-20230823-1820
Updated: 2023-08-23T18:20:33Z\nUpdated Date: 1692814833\nNow: 1693248938\nDiff: 434105
Checking tag: 71b7ac2-20230823-1231
Updated: 2023-08-23T12:32:03Z\nUpdated Date: 1692793923\nNow: 1693248938\nDiff: 455015
Checking tag: 71b7ac2-20230823-0621
Updated: 2023-08-23T06:21:58Z\nUpdated Date: 1692771718\nNow: 1693248938\nDiff: 477220
Checking tag: 71b7ac2-20230823-0103
Updated: 2023-08-23T01:04:13Z\nUpdated Date: 1692752653\nNow: 1693248938\nDiff: 496285
Checking tag: 71b7ac2-20230822-1851
Updated: 2023-08-22T18:51:26Z\nUpdated Date: 1692730286\nNow: 1693248938\nDiff: 518652
Checking tag: 71b7ac2-20230822-1230
Updated: 2023-08-22T12:31:26Z\nUpdated Date: 1692707486\nNow: 1693248938\nDiff: 541452
Checking tag: 71b7ac2-20230822-0622
Updated: 2023-08-22T06:22:45Z\nUpdated Date: 1692685365\nNow: 1693248938\nDiff: 563573
Checking tag: 71b7ac2-20230822-0104
Updated: 2023-08-22T01:05:15Z\nUpdated Date: 1692666315\nNow: 1693248938\nDiff: 582623
Checking tag: 87d1057-20230821-1933
Updated: 2023-08-21T19:34:01Z\nUpdated Date: 1692646441\nNow: 1693248938\nDiff: 602497
Found 21 tags to delete.
Deleting tag: e6f0942-20230826-1819
Deleting tag: e6f0942-20230826-1227
Deleting tag: 96a3e4c-20230826-0620
Deleting tag: 71b7ac2-20230826-0103
Deleting tag: 71b7ac2-20230825-1821
Deleting tag: 71b7ac2-20230825-1231
Deleting tag: 71b7ac2-20230825-0621
Deleting tag: 71b7ac2-20230825-0105
Deleting tag: 71b7ac2-20230824-1821
Deleting tag: 71b7ac2-20230824-1231
Deleting tag: 71b7ac2-20230824-0621
Deleting tag: 71b7ac2-20230824-0104
Deleting tag: 71b7ac2-20230823-1820
Deleting tag: 71b7ac2-20230823-1231
Deleting tag: 71b7ac2-20230823-0621
Deleting tag: 71b7ac2-20230823-0103
Deleting tag: 71b7ac2-20230822-1851
Deleting tag: 71b7ac2-20230822-1230
Deleting tag: 71b7ac2-20230822-0622
Deleting tag: 71b7ac2-20230822-0104
Deleting tag: 87d1057-20230821-1933
```

The output looks correct I think.

raisedadead commented:
Please feel free to run this live. I still need to start using the images from DOCR, so we have some leeway if you want to experiment.

naomi-lgbt commented:
Cool, I'll fire a run on my [test repo](https://github.com/naomi-lgbt/actions-testing/blob/main/.github/workflows/deploy-i18n.yml) to specifically look at the clean up steps.

Commit:
chore: pass tag correctly

Commit:
chore: use the force Naomi

naomi-lgbt commented:
Okay we have a run which [successfully deleted the tags](https://github.com/naomi-lgbt/actions-testing/actions/runs/6005155363), and a subsequent run which confirms [garbage collection is triggered](https://github.com/naomi-lgbt/actions-testing/actions/runs/6005193002/job/16287408600).

naomi-lgbt commented:
Aaaand the garbage collection successfully removed the untagged images as expected. 🎉

Commit:
chore: use tld from matrix
